docs/how-to/model-debug.txt

```diff
@@ -0,0 +1,337 @@
+      a message saying ``Checkpoint storage validation failed``, please
+      review the correctness of the values in your
+      ``checkpoint_storage`` settings.
+
+   -  The experiment config is fully validated for cluster-managed
+      experiments, more strictly than it is for ``--local --test`` mode.
+      If you get errors related to ``unmarshaling JSON`` when trying to
+      submit the experiment to the cluster, that is an indication that
+      the experiment config has errors. Please review the TODO:
+      experiment-config-reference
+
+Again, if you are unable to identify the root cause of the issue yourself,
+please do not hesitate to contact Determined through our community
+support channels! TODO: link to support
+
+8. Does a single-GPU experiment work?
+-------------------------------------
+
+This step is just like to Step 7, except it introduces hyperparameter
+search and will execute full training for each trial.
+
+**How to test:** Configuration should be identical to Step 7.  Again,
+confirm that your experiment config either does not specify
+``resources.slots_per_trial`` at all, or that it is set to 1, like:
+
+.. code:: yaml
+
+   resources:
+     slots_per_trial: 1
+
+Then create an experiment without either of ``--test`` or ``--local``
+flags (probably you will find the ``--follow`` or ``-f`` flag to be
+helpful):
+
+.. code:: bash
+
+   det experiment create myconfig.yaml my_model_dir -f
+
+**How to diagnose failures:** If Step 7 worked but Step 8 does not,
+there are a few high-level categories of issues to check for:
+
+   -  Does the error happen when the experiment has a
+      ``searcher.source_trial_id`` set?  One thing that can occur in a
+      real experiment that does not occur in a ``--test`` experiment is
+      the loading of a of a previous checkpoint.  Errors when loading
+      from a checkpoint can be caused by architecture change, where the
+      new model code is not architecture-compatible with the old trial
+      code.
+
+   - Generally, issues at this step are caused by doing training and
+     evaluation continuously, so focus on how that change may cause
+     issues with your code.
+
+9. Does a multi-GPU experiment work?
+-------------------------------------
+
+This step is like Step 8, but introduces distributed training.
+Naturally, this step is only relevant if you have multiple GPUs and you
+wish to use distributed training.
+
+**How to test:** Configuration should be like Step 7, except you will
+now set ``resources.slots_per_trial`` to some number greater than 1:
+
+.. code:: yaml
+
+   resources:
+     slots_per_trial: 2
+
+Then create your experiment:
+
+.. code:: bash
+
+   det experiment create myconfig.yaml my_model_dir -f
+
+**How to diagnose failures:**  If you are using the ``determined``
+library APIs correctly, then theoretically distributed training should
+"just work".  However, you should be aware of some common pitfalls:
```

Comment:
Talk about networking issues? And maybe scheduling problems / hangs?

Reply:
Added a detail about not being scheduled.

Are there common networking issue that affect distributed training and not normal training?   Almost all of the networking issues I've helped debug were issues that affected even a single notebook.

re hangs: I think hangs in our harness are nearly 100% our responsibility.  I can't think of any hangs that were due to the user doing something wrong.

Reply:
Yeah -- I mostly meant hangs due to dtrain trials not being scheduled.

Re: networking, I was thinking of issues like: (a) matching network interface names (b) additional ports we use for NCCL, ssh, etc. traffic we don't use for single-GPU training (c) issues with istio or other proxies that seem to be more problematic for dtrain trials (e.g., https://determined-community.slack.com/archives/CV3MTNZ6U/p1607112511338900)

Reply:
I think things like port configurations belong in a separate "debug/validate your determined cluster", where we would have a series of tests that new cluster admins could run to ensure that their users' code would work.  I think this document should target the ml end user